File: js/packages/teams-ai/src/augmentations/DefaultAugmentation.ts
/**
 * @module teams-ai
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import { TurnContext } from 'botbuilder-core';
import { PromptResponse } from '../models';
import { Plan, PredictedSayCommand } from '../planners';
import { Tokenizer } from '../tokenizers';
import { Validation } from '../validators';
import { Augmentation } from './Augmentation';
import { PromptSection } from '../prompts';
import { Memory } from '../MemoryFork';

/**
 * The default 'none' augmentation.
 * @remarks
 * This augmentation does not add any additional functionality to the prompt. It always
 * returns a `Plan` with a single `SAY` command containing the models response.
 */
export class DefaultAugmentation implements Augmentation<string> {
    /**
     * @returns {PromptSection|undefined} Returns an optional prompt section for the augmentation.
     */
    public createPromptSection(): PromptSection | undefined {
        return undefined;
    }

    /**
     * Validates a response to a prompt.
     * @param {TurnContext} context - Context for the current turn of conversation with the user.
     * @param {Memory} memory - An interface for accessing state values.
     * @param {Tokenizer} tokenizer - Tokenizer to use for encoding and decoding text.
     * @param {PromptResponse<string>} response - Response to validate.
     * @param {number} remaining_attempts Number of remaining attempts to validate the response.
     * @returns {Validation} A `Validation` object.
     */
    public validateResponse(
        context: TurnContext,
        memory: Memory,
        tokenizer: Tokenizer,
        response: PromptResponse<string>,
        remaining_attempts: number
    ): Promise<Validation<string>> {
        return Promise.resolve({
            type: 'Validation',
            valid: true
        });
    }

    /**
     * Creates a plan given validated response value.
     * @param {TurnContext} context Context for the current turn of conversation.
     * @param {Memory} memory An interface for accessing state variables.
     * @param {PromptResponse<string>} response The validated and transformed response for the prompt.
     * @returns {Promise<Plan>} The created plan.
     */
    public createPlanFromResponse(
        context: TurnContext,
        memory: Memory,
        response: PromptResponse<string>
    ): Promise<Plan> {
        return Promise.resolve({
            type: 'plan',
            commands: [
                {
                    type: 'SAY',
                    response: response.message
                } as PredictedSayCommand
            ]
        });
    }
}

File: python/packages/ai/teams/ai/augmentations/default_augmentation.py
"""
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the MIT License.
"""

from __future__ import annotations

from typing import Union

from botbuilder.core import TurnContext

from ...state import MemoryBase
from ..models.prompt_response import PromptResponse
from ..planners.plan import Plan, PredictedSayCommand
from ..prompts.sections.prompt_section import PromptSection
from ..tokenizers.tokenizer import Tokenizer
from ..validators.validation import Validation
from .augmentation import Augmentation


class DefaultAugmentation(Augmentation[str]):
    """
    The default 'none' augmentation.

    This augmentation does not add any additional functionality to the prompt. It always
    returns a `Plan` with a single `SAY` command containing the models response.
    """

    def create_prompt_section(self) -> Union[PromptSection, None]:
        """
        Creates an optional prompt section for the augmentation.
        """
        return None

    async def validate_response(
        self,
        context: TurnContext,
        memory: MemoryBase,
        tokenizer: Tokenizer,
        response: PromptResponse[str],
        remaining_attempts: int,
    ) -> Validation:
        """
        Validates a response to a prompt.

        Args:
            context (TurnContext): Context for the current turn of conversation.
            memory (MemoryBase): Interface for accessing state variables.
            tokenizer (Tokenizer): Tokenizer to use for encoding/decoding text.
            response (PromptResponse[str]): Response to validate.
            remaining_attempts (int): Nubmer of remaining attempts to validate the response.

        Returns:
            Validation: A 'Validation' object.
        """
        return Validation(valid=True)

    async def create_plan_from_response(
        self, turn_context: TurnContext, memory: MemoryBase, response: PromptResponse[str]
    ) -> Plan:
        """
        Creates a plan given validated response value.

        Args:
            turn_context (TurnContext): Context for the current turn of conversation.
            memory (MemoryBase):  Interface for accessing state variables.
            response (PromptResponse[str]): The validated and transformed response for the prompt.

        Returns:
            Plan: The created plan.
        """
        say_response = ""
        if response.message and response.message.content:
            say_response = response.message.content
        return Plan(commands=[PredictedSayCommand(response=say_response)])
